README.md
# TSLint Rules
Zbiór reguł `tslint` wykorzystywanych w projektach OPI.

## Instalacja paczki
W głównym katalogu projektu (tj. tam gdzie znajduje się plik `package.json`) uruchom:

``
npm i @opi/tslint-rules -D
``

## Konfiguracja
W konfiguracji `tslint` (domyślnie plik `tslint.json`) dodaj wpis w sekcji `rulesDirectory`:
```
{
  ...
  "rulesDirectory": [
    ...
    "./node_modules/@opi/tslint-rules/dist"
    ...
  ],
  ...
}
```

## Opis reguł
| Reguła | Opis |
|-------------------------|------------------|
| `consistent-decorators` | Uniemożliwia wstawianie dekoratorów w tej samej linii co dekorowane wyrażenie (nie dotyczy przypadku kiedy dekorator używany jest dla parametru funkcji / konstruktora -- wówczas dekorator i argument muszą być w tej samej linii).
| `ng-consistent-naming`  | Pilnuje spójności w nazewnictwie bytów Angularowych według reguł opisanych na [https://angular.io/guide/styleguide#symbols-and-file-names](https://angular.io/guide/styleguide#symbols-and-file-names). Jako parametr reguła przyjmuje obiekt postaci `{"vendorPrefixes": prefixes}`, gdzie `prefixes` to tablica prefiksów dla selektorów komponentów/dyrektyw używanych w danych projekcie.   
| `no-empty-lines-near-brackets`  | Uniemożliwia pozostawianie pustych linii po lub przed nawiasem / klamrą (nie dotyczy klamer otwierających bloki instrukcji). |
| `require-license-banner` | Sprawdza czy treść pliku zawiera na początku odpowiedni komentarz z licencją. Jako parametr reguła przyjmuje obiekt postaci: `{ "bannerFile": file }` gdzie `file` to nazwa pliku w którym zawarta jest treść komentarza z licencją.


CHANGELOG.md
# 1.1.0 (2018-05-30)
### Nowe funkcjonalności
* Nowa reguła: `ng-consistent-naming`

# 1.0.1 (2018-05-14)

### Poprawki błędów
* `require-license-banner`: prawidłowe rozwiązywanie ścieżki z plikiem licencyjnym

# 1.0.0 (2018-05-14) 

### Wydanie początkowe

Pierwsze reguły:
* `consistent-decorators`
* `no-empty-lines-near-brackets`
* `require-license-banner`


test/consistentDecoratorsRule.bad.ts
/**
 * @license
 *
 * Example license banner
 */
function XYZ() {
}

@XYZ() class AClass {

    @XYZ() aField: string;

    constructor(@XYZ()
                private _bField: string) {
    }

    method(@XYZ()
           a: string) {
    }
}


test/ng-consistent-named-bad.pipe.ts
/**
 * @license
 *
 * Example license banner
 */
import {Pipe} from "@angular/core";

@Pipe({
    name: "ngConsistentNamedBad",
})
export class NgConsistentNamedBadBadPipe {
}


test/requireLicenseBannerRule.bad.ts
/**
 * @license
 *
 * Example license banner FOOOO
 */

/* ^ */
/* | */
/* | */
/* | */
/* This is BAD license banner */


src/consistentDecoratorsRule.ts
/**
 * @license
 * To oprogramowanie jest własnością
 *
 * OPI - Ośrodek Przetwarzania Informacji,
 * Al. Niepodległości 188B, 00-608 Warszawa
 * Numer KRS: 0000127372
 * Sąd Rejonowy dla m. st. Warszawy w Warszawie XII Wydział
 * Gospodarczy KRS
 * REGON: 006746090
 * NIP: 525-000-91-40
 *
 * Wszystkie prawa zastrzeżone. To oprogramowanie może być używane tylko
 * zgodnie z przeznaczeniem. OPI nie odpowiada za ewentualne wadliwe
 * działanie kodu.
 */
import {AbstractWalker, RuleFailure, Rules} from "tslint";
import {isDecorator, isSameLine, isConstructorDeclaration, isMethodDeclaration} from "tsutils";
import {forEachChild, Node, SourceFile} from "typescript";

function isDecoratorInConstructorOrMethod(node: Node): boolean {
    return (
        node && node.parent && node.parent.parent &&
        (isMethodDeclaration(node.parent.parent) || isConstructorDeclaration(node.parent.parent))
    );
}

class ConsistentDecoratorsWalker extends AbstractWalker<any> {

    walk(sourceFile: SourceFile): void {
        const callback: any = (node: Node): void => {
            if (isDecorator(node) && node.parent) {
                const decoratedExpressionEnd: number = node.parent.getChildAt(1, this.sourceFile).end;

                if (isDecoratorInConstructorOrMethod(node)) {
                    if (!isSameLine(this.sourceFile, node.end, decoratedExpressionEnd)) {
                        this.addFailure(
                            node.end, node.end,
                            "Dekorowane wyrażenie musi być w tej samej linii co dekorator",
                        );
                    }
                } else {
                    if (isSameLine(this.sourceFile, node.end, decoratedExpressionEnd)) {
                        this.addFailure(
                            node.end, node.end,
                            "Dekorowane wyrażenie nie może być w tej samej linii co dekorator",
                        );
                    }
                }
            }

            return forEachChild(node, callback);
        };

        return forEachChild(sourceFile, callback);
    }
}

export class Rule extends Rules.AbstractRule {

    apply(sourceFile: SourceFile): RuleFailure[] {
        return this.applyWithWalker(new ConsistentDecoratorsWalker(sourceFile, this.ruleName, {}));
    }
}


src/requireLicenseBannerRule.ts
/**
 * @license
 * To oprogramowanie jest własnością
 *
 * OPI - Ośrodek Przetwarzania Informacji,
 * Al. Niepodległości 188B, 00-608 Warszawa
 * Numer KRS: 0000127372
 * Sąd Rejonowy dla m. st. Warszawy w Warszawie XII Wydział
 * Gospodarczy KRS
 * REGON: 006746090
 * NIP: 525-000-91-40
 *
 * Wszystkie prawa zastrzeżone. To oprogramowanie może być używane tylko
 * zgodnie z przeznaczeniem. OPI nie odpowiada za ewentualne wadliwe
 * działanie kodu.
 */

/**
 * @see https://github.com/angular/material2/blob/master/tools/tslint-rules/requireLicenseBannerRule.ts
 */
import {readFileSync} from "fs";
import {resolve, dirname} from "path";
import {IOptions, Replacement, RuleFailure, Rules, RuleWalker} from "tslint";
import {SourceFile} from "typescript";

interface RequireLicenseBannerRuleOptions {
    bannerFile: string;
}

interface RequireLicenseBannerWalkerConfig extends RequireLicenseBannerRuleOptions {
    bannerContent: string;
}

function getTsLintConfigDir(): string {
    const tslintConfigArgvIndex: number = process.argv.findIndex(arg => arg === "-c" || arg === "--config");
    const tslintConfigFile: string = tslintConfigArgvIndex > -1 ? process.argv[tslintConfigArgvIndex + 1] : "./tsconfig.json";

    return dirname(resolve(tslintConfigFile));
}

class RequireLicenseBannerWalker extends RuleWalker {

    private _tslintFix: Replacement;

    constructor(sourceFile: SourceFile, options: IOptions, private _config: RequireLicenseBannerWalkerConfig) {
        super(sourceFile, options);
        this._tslintFix = Replacement.appendText(0, `${this._config.bannerContent}`);
    }

    visitSourceFile(sourceFile: SourceFile): void {
        const fileContent: string = sourceFile.getFullText().replace(/\r\n/g, "\n");
        const bannerCommentPos: number = fileContent.indexOf(this._config.bannerContent);

        if (bannerCommentPos !== 0) {
            const error: string = `Zawartość pliku musi rozpoczynać się komentarzem z licencją` +
                ` (treść można przekleić z pliku ${this._config.bannerFile})`;
            return this.addFailureAt(0, 0, error, this._tslintFix);
        }

        super.visitSourceFile(sourceFile);
    }
}

export class Rule extends Rules.AbstractRule {

    private _walkerConfig: RequireLicenseBannerWalkerConfig;

    constructor(options: IOptions) {
        super(options);
        if (options.ruleArguments && options.ruleArguments.length > 0) {
            const ruleOptions: RequireLicenseBannerRuleOptions = options.ruleArguments[0];
            const bannerFile: string = resolve(getTsLintConfigDir(), ruleOptions.bannerFile);

            this._walkerConfig = {
                bannerFile,
                bannerContent: readFileSync(bannerFile).toString().replace(/\r\n/g, "\n"),
            };
        }
    }

    apply(sourceFile: SourceFile): RuleFailure[] {
        return this.applyWithWalker(new RequireLicenseBannerWalker(sourceFile, this.getOptions(), this._walkerConfig));
    }
}


src/ngConsistentFilenamesRule.ts
/**
 * @license
 * To oprogramowanie jest własnością
 *
 * OPI - Ośrodek Przetwarzania Informacji,
 * Al. Niepodległości 188B, 00-608 Warszawa
 * Numer KRS: 0000127372
 * Sąd Rejonowy dla m. st. Warszawy w Warszawie XII Wydział
 * Gospodarczy KRS
 * REGON: 006746090
 * NIP: 525-000-91-40
 *
 * Wszystkie prawa zastrzeżone. To oprogramowanie może być używane tylko
 * zgodnie z przeznaczeniem. OPI nie odpowiada za ewentualne wadliwe
 * działanie kodu.
 */
import {basename} from "path";
import {AbstractWalker, RuleFailure, Rules} from "tslint";
import {
    isCallExpression,
    isClassDeclaration,
    isIdentifier,
    isObjectLiteralExpression,
    isPropertyAssignment,
    isStringLiteral,
} from "tsutils";
import {forEachChild, CallExpression, Expression, Node, SourceFile} from "typescript";

const STYLEGUIDE_URL: string = "https://angular.io/guide/styleguide#symbols-and-file-names";

function dashCaseToCamelCase(str: string): string {
    return str.replace(/-([a-z])/g, g => g[1].toUpperCase());
}

function getObjectLiteralFromExpression(expression: Expression): any {
    if (expression && isObjectLiteralExpression(expression)) {
        const result: any = {};

        for (const property of expression.properties) {
            if (isPropertyAssignment(property) && isIdentifier(property.name) && isStringLiteral(property.initializer)) {
                result[property.name.escapedText.toString()] = property.initializer.text.toString();
            }
        }

        return result;
    }
    return null;
}

function isConsistentNgNaming(className: string, classSuffix: string, expression: string,
                              matcher: (_ngClassName: string, _expression: string) => boolean): boolean {
    if (typeof expression === "string") {
        const ngClassName: string = className.replace(new RegExp(`${classSuffix}$`), "");

        return matcher(ngClassName, expression);
    }
    return true;
}

function isConsistentFileNaming(fileName: string, className: string, classSuffix: string, fileNameSuffix: string): boolean {
    const ngFileName: string = basename(fileName, fileNameSuffix);
    const ngClassName: string = className.replace(new RegExp(`${classSuffix}$`), "");

    return dashCaseToCamelCase(`-${ngFileName}`) === ngClassName;
}

class NgConsistentFilenamesWalker extends AbstractWalker<any> {

    walk(sourceFile: SourceFile): void {
        const callback: any = (node: Node): void => {
            if (isClassDeclaration(node) && node.decorators && node.decorators.length > 0) {
                for (const decorator of node.decorators) {
                    if (isCallExpression(decorator.expression)) {
                        const decoratorCallExpression: CallExpression = decorator.expression;

                        if (isIdentifier(decoratorCallExpression.expression)) {
                            const className: string = node.name.escapedText.toString();
                            const decoratorName: string = decoratorCallExpression.expression.escapedText.toString();
                            const decoratorArgs: any = getObjectLiteralFromExpression(decoratorCallExpression.arguments[0]);

                            if (decoratorName === "NgModule") {
                                this._validateNgModuleNaming(className);
                            } else if (decoratorName === "Injectable") {
                                this._validateNgServiceNaming(className);
                            } else if (decoratorName === "Component") {
                                this._validateNgComponentNaming(className, decoratorArgs);
                            } else if (decoratorName === "Directive") {
                                this._validateNgDirectiveNaming(className, decoratorArgs);
                            } else if (decoratorName === "Pipe") {
                                this._validateNgPipeNaming(className, decoratorArgs);
                            }
                        }
                    }
                }
            }

            return forEachChild(node, callback);
        };

        return forEachChild(sourceFile, callback);
    }

    private _validateNgModuleNaming(className: string): void {
        if (!isConsistentFileNaming(this.sourceFile.fileName, className, "Module", ".module.ts")) {
            this._addFailure("Niespójność w nazewnictwie modułu Angularowego");
        }
    }

    private _validateNgServiceNaming(className: string): void {
        if (!isConsistentFileNaming(this.sourceFile.fileName, className, "Service", ".service.ts")) {
            this._addFailure("Niespójność w nazewnictwie serwisu Angularowego");
        }
    }

    private _validateNgComponentNaming(className: string, params: any): void {
        if (!isConsistentFileNaming(this.sourceFile.fileName, className, "Component", ".component.ts")) {
            this._addFailure("Niespójność w nazewnictwie komponentu Angularowego");
        }
        if (!isConsistentFileNaming(params["templateUrl"], className, "Component", ".component.html")) {
            this._addFailure("Niespójność w nazewnictwie szablonu komponentu Angularowego");
        }
        if (!isConsistentNgNaming(className, "Component", params["selector"],
                (ngClassName, selector) => ngClassName === dashCaseToCamelCase(selector.replace(/^test/, "")))) {
            this._addFailure("Selektor niezgodny z nazwą klasy komponentu Angularowego");
        }
    }

    private _validateNgDirectiveNaming(className: string, params: any): void {
        if (!isConsistentFileNaming(this.sourceFile.fileName, className, "Directive", ".directive.ts")) {
            this._addFailure("Niespójność w nazewnictwie dyrektywy Angularowej");
        }
        if (!isConsistentNgNaming(className, "Directive", params["selector"],
                (ngClassName, selector) => selector.match(new RegExp(`\\[test${ngClassName}\\]`)) !== null)) {
            this._addFailure("Selektor niezgodny z nazwą klasy dyrektywy Angularowej");
        }
    }

    private _validateNgPipeNaming(className: string, params: any): void {
        if (!isConsistentFileNaming(this.sourceFile.fileName, className, "Pipe", ".pipe.ts")) {
            this._addFailure("Niespójność w nazewnictwie pipe'a Angularowego");
        }
        if (!isConsistentNgNaming(className, "Pipe", params["name"],
                (ngClassName, name) => ngClassName === name.replace(/^(.)/, match => match.toUpperCase()))) {
            this._addFailure("Nazwa niezgodna z nazwą klasy pipe'a Angularowego");
        }
    }

    private _addFailure(message: string): void {
        this.addFailure(0, 0, `${message} (patrz: ${STYLEGUIDE_URL})`);
    }
}

export class Rule extends Rules.AbstractRule {

    apply(sourceFile: SourceFile): RuleFailure[] {
        return this.applyWithWalker(new NgConsistentFilenamesWalker(sourceFile, this.ruleName, {}));
    }
}


src/noSingleLineDecoratorsRule.ts
/**
 * @license
 * To oprogramowanie jest własnością
 *
 * OPI - Ośrodek Przetwarzania Informacji,
 * Al. Niepodległości 188B, 00-608 Warszawa
 * Numer KRS: 0000127372
 * Sąd Rejonowy dla m. st. Warszawy w Warszawie XII Wydział
 * Gospodarczy KRS
 * REGON: 006746090
 * NIP: 525-000-91-40
 *
 * Wszystkie prawa zastrzeżone. To oprogramowanie może być używane tylko
 * zgodnie z przeznaczeniem. OPI nie odpowiada za ewentualne wadliwe
 * działanie kodu.
 */
import {AbstractWalker, RuleFailure, Rules} from "tslint";
import {isDecorator, isSameLine} from "tsutils";
import {forEachChild, Node, SourceFile, SyntaxKind} from "typescript";

function isDecoratorInConstructor(node: Node): boolean {
    return (node && node.parent && node.parent.parent && node.parent.parent.kind === SyntaxKind.Constructor);
}

class NoSingleLineDecoratorsWalker extends AbstractWalker<any> {

    walk(sourceFile: SourceFile): void {
        const callback: any = (node: Node): void => {
            if (isDecorator(node) && node.parent) {
                const decoratedExpressionEnd: number = node.parent.getChildAt(1, this.sourceFile).end;

                if (!isDecoratorInConstructor(node) && isSameLine(this.sourceFile, node.end, decoratedExpressionEnd)) {
                    this.addFailure(
                        node.end, node.end,
                        "Dekorowane wyrażenie nie może być w tej samej linii co dekorator",
                    );
                }
            }

            return forEachChild(node, callback);
        };

        return forEachChild(sourceFile, callback);
    }
}

export class Rule extends Rules.AbstractRule {

    apply(sourceFile: SourceFile): RuleFailure[] {
        return this.applyWithWalker(new NoSingleLineDecoratorsWalker(sourceFile, this.ruleName, {}));
    }
}


src/noEmptyLinesNearBracketsRule.ts
/**
 * @license
 * To oprogramowanie jest własnością
 *
 * OPI - Ośrodek Przetwarzania Informacji,
 * Al. Niepodległości 188B, 00-608 Warszawa
 * Numer KRS: 0000127372
 * Sąd Rejonowy dla m. st. Warszawy w Warszawie XII Wydział
 * Gospodarczy KRS
 * REGON: 006746090
 * NIP: 525-000-91-40
 *
 * Wszystkie prawa zastrzeżone. To oprogramowanie może być używane tylko
 * zgodnie z przeznaczeniem. OPI nie odpowiada za ewentualne wadliwe
 * działanie kodu.
 */
import {AbstractWalker, RuleFailure, Rules} from "tslint";
import {
    isArrayLiteralExpression,
    isBlock,
    isCallExpression,
    isClassDeclaration,
    isInterfaceDeclaration,
    isObjectLiteralExpression,
    isParenthesizedExpression,
    isSameLine,
    isTypeAliasDeclaration,
} from "tsutils";
import {forEachChild, Node, SourceFile} from "typescript";

function isBracketExpression(node: Node): boolean {
    return isBlock(node) ||
        isParenthesizedExpression(node) ||
        isClassDeclaration(node) ||
        isInterfaceDeclaration(node) ||
        isTypeAliasDeclaration(node) ||
        isArrayLiteralExpression(node) ||
        isObjectLiteralExpression(node) ||
        isCallExpression(node);
}

function canHaveOpeningBlankLine(node: Node): boolean {
    return isBlock(node) ||
        isClassDeclaration(node) ||
        isInterfaceDeclaration(node);
}

class NoEmptyLinesNearBracketsWalker extends AbstractWalker<any> {

    walk(sourceFile: SourceFile): void {
        const callback: any = (node: Node): void => {
            if (isBracketExpression(node) && !isSameLine(this.sourceFile, node.pos, node.end)) {
                const expressionLines: string[] = node.getText(this.sourceFile).split("\n");

                if (expressionLines.length > 2) {
                    if (expressionLines[1].trim() === "" && !canHaveOpeningBlankLine(node)) {
                        this.addFailure(
                            node.end - node.getText(this.sourceFile).length, node.end,
                            "Niepotrzebna pusta linia po nawiasie otwierającym",
                        );
                    }
                    if (expressionLines[expressionLines.length - 2].trim() === "") {
                        this.addFailure(
                            node.end, node.end,
                            "Niepotrzebna pusta linia przed nawiasem zamykającym",
                        );
                    }
                }
            }

            return forEachChild(node, callback);
        };

        return forEachChild(sourceFile, callback);
    }
}

export class Rule extends Rules.AbstractRule {

    apply(sourceFile: SourceFile): RuleFailure[] {
        return this.applyWithWalker(new NoEmptyLinesNearBracketsWalker(sourceFile, this.ruleName, {}));
    }
}


src/ngConsistentNamingRule.ts
/**
 * @license
 * To oprogramowanie jest własnością
 *
 * OPI - Ośrodek Przetwarzania Informacji,
 * Al. Niepodległości 188B, 00-608 Warszawa
 * Numer KRS: 0000127372
 * Sąd Rejonowy dla m. st. Warszawy w Warszawie XII Wydział
 * Gospodarczy KRS
 * REGON: 006746090
 * NIP: 525-000-91-40
 *
 * Wszystkie prawa zastrzeżone. To oprogramowanie może być używane tylko
 * zgodnie z przeznaczeniem. OPI nie odpowiada za ewentualne wadliwe
 * działanie kodu.
 */
import {basename} from "path";
import {AbstractWalker, IOptions, RuleFailure, Rules} from "tslint";
import {
    isCallExpression,
    isClassDeclaration,
    isIdentifier,
    isObjectLiteralExpression,
    isPropertyAssignment,
    isStringLiteral,
} from "tsutils";
import {forEachChild, CallExpression, Expression, Node, SourceFile} from "typescript";

const STYLEGUIDE_URL: string = "https://angular.io/guide/styleguide#symbols-and-file-names";

interface NgConsistentNamingRuleOptions {
    vendorPrefixes: string[];
}

function isDashCased(str: string): boolean {
    return typeof str === "string" && str.match(/^[a-z\-]+$/g) !== null;
}

function dashCaseToCamelCase(str: string): string {
    if (!isDashCased(str)) {
        return null;
    }
    return str.replace(/-([a-z])/g, g => g[1].toUpperCase());
}

function getObjectLiteralFromExpression(expression: Expression): any {
    if (expression && isObjectLiteralExpression(expression)) {
        const result: any = {};

        for (const property of expression.properties) {
            if (isPropertyAssignment(property) && isIdentifier(property.name) && isStringLiteral(property.initializer)) {
                result[property.name.escapedText.toString()] = property.initializer.text.toString();
            }
        }

        return result;
    }
    return null;
}

function isConsistentNgNaming(className: string, classSuffix: string, expression: string,
                              matcher: (_ngClassName: string, _expression: string) => boolean): boolean {
    if (typeof expression === "string") {
        const ngClassName: string = className.replace(new RegExp(`${classSuffix}$`), "");

        return matcher(ngClassName, expression);
    }
    return true;
}

function isConsistentFileNaming(fileName: string, className: string, classSuffix: string, fileNameSuffix: string): boolean {
    const ngFileName: string = basename(fileName, fileNameSuffix);
    const ngClassName: string = className.replace(new RegExp(`${classSuffix}$`), "");

    return dashCaseToCamelCase(`-${ngFileName}`) === ngClassName;
}

class NgConsistentNamingWalker extends AbstractWalker<NgConsistentNamingRuleOptions> {

    walk(sourceFile: SourceFile): void {
        const callback: any = (node: Node): void => {
            if (isClassDeclaration(node) && node.decorators && node.decorators.length > 0) {
                for (const decorator of node.decorators) {
                    if (isCallExpression(decorator.expression)) {
                        const decoratorCallExpression: CallExpression = decorator.expression;

                        if (isIdentifier(decoratorCallExpression.expression)) {
                            const className: string = node.name.escapedText.toString();
                            const decoratorName: string = decoratorCallExpression.expression.escapedText.toString();
                            const decoratorArgs: any = getObjectLiteralFromExpression(decoratorCallExpression.arguments[0]);

                            if (decoratorName === "NgModule") {
                                this._validateNgModuleNaming(className);
                            } else if (decoratorName === "Injectable") {
                                this._validateNgServiceNaming(className);
                            } else if (decoratorName === "Component") {
                                this._validateNgComponentNaming(className, decoratorArgs);
                            } else if (decoratorName === "Directive") {
                                this._validateNgDirectiveNaming(className, decoratorArgs);
                            } else if (decoratorName === "Pipe") {
                                this._validateNgPipeNaming(className, decoratorArgs);
                            }
                        }
                    }
                }
            }

            return forEachChild(node, callback);
        };

        return forEachChild(sourceFile, callback);
    }

    private _validateNgModuleNaming(className: string): void {
        if (!isConsistentFileNaming(this.sourceFile.fileName, className, "Module", ".module.ts")) {
            this._addFailure("Niespójność w nazewnictwie modułu Angularowego");
        }
    }

    private _validateNgServiceNaming(className: string): void {
        if (!isConsistentFileNaming(this.sourceFile.fileName, className, "Service", ".service.ts")) {
            this._addFailure("Niespójność w nazewnictwie serwisu Angularowego");
        }
    }

    private _validateNgComponentNaming(className: string, params: any): void {
        const vendorSelectorRegex: RegExp = new RegExp(`^(${this.options.vendorPrefixes.join("|")})`);

        if (!isConsistentFileNaming(this.sourceFile.fileName, className, "Component", ".component.ts")) {
            this._addFailure("Niespójność w nazewnictwie komponentu Angularowego");
        }
        if (!isConsistentFileNaming(params["templateUrl"], className, "Component", ".component.html")) {
            this._addFailure("Niespójność w nazewnictwie szablonu komponentu Angularowego");
        }
        if (!isConsistentNgNaming(className, "Component", params["selector"],
                (ngClassName, selector) => ngClassName === dashCaseToCamelCase(selector.replace(vendorSelectorRegex, "")))) {
            this._addFailure("Selektor niezgodny z nazwą klasy komponentu Angularowego");
        }
    }

    private _validateNgDirectiveNaming(className: string, params: any): void {
        const vendorPrefixes: string = this.options.vendorPrefixes.join("|");

        if (!isConsistentFileNaming(this.sourceFile.fileName, className, "Directive", ".directive.ts")) {
            this._addFailure("Niespójność w nazewnictwie dyrektywy Angularowej");
        }
        if (!isConsistentNgNaming(className, "Directive", params["selector"],
                (ngClassName, selector) => selector.match(new RegExp(`\\[${vendorPrefixes}${ngClassName}\\]`)) !== null)) {
            this._addFailure("Selektor niezgodny z nazwą klasy dyrektywy Angularowej");
        }
    }

    private _validateNgPipeNaming(className: string, params: any): void {
        if (!isConsistentFileNaming(this.sourceFile.fileName, className, "Pipe", ".pipe.ts")) {
            this._addFailure("Niespójność w nazewnictwie pipe'a Angularowego");
        }
        if (!isConsistentNgNaming(className, "Pipe", params["name"],
                (ngClassName, name) => ngClassName === name.replace(/^(.)/, match => match.toUpperCase()))) {
            this._addFailure("Nazwa niezgodna z nazwą klasy pipe'a Angularowego");
        }
    }

    private _addFailure(message: string): void {
        this.addFailure(0, 0, `${message} (patrz: ${STYLEGUIDE_URL})`);
    }
}

export class Rule extends Rules.AbstractRule {

    private _walkerOptions: NgConsistentNamingRuleOptions = { vendorPrefixes: [] };

    constructor(options: IOptions) {
        super(options);
        if (options.ruleArguments && options.ruleArguments.length > 0) {
            this._walkerOptions = options.ruleArguments[0];
        }
    }

    apply(sourceFile: SourceFile): RuleFailure[] {
        return this.applyWithWalker(new NgConsistentNamingWalker(sourceFile, this.ruleName, this._walkerOptions));
    }
}
